test/test_security_service.py

```diff
@@ -204,9 +204,9 @@ def test_vm_firewall_group_rule(self):
                 r.delete()
             fw = self.provider.security.vm_firewalls.get(fw.id)  # update
             self.assertTrue(
-                len(list(fw.rules)) == 0,
+                fw is None or len(list(fw.rules)) == 0,
```

Comment:
It is already the case that we create a dummy rule when we create a firewall, which simplified a lot of other tests. But, this particular test is explicitly deleting all rules, including the dummy rule that was created (lines 203 and 204). That's why vm_firewalls.get may return None.

Reply:
Right, in the other cloud providers, a firewall with no rules will continue to exist. Can we change GCE behaviour to match so that deleting all the rules will leave the dummy rule intact, so that the firewall will continue to exist? This will make the behaviour consistent with other providers. Calling firewall.delete() can delete the remaining dummy rule, permanently removing the firewall from existence.

Reply:
Let me try to understand the suggestion: are you saying that calling r.delete() on the dummy rule should not delete it? If so, then this test should assert that len(list(fw.rules)) == 1 not 0, because the dummy rule will still exist. Maybe you think fw.rules should not return the dummy rule? Then, that contradicts the test a few lines about this test that asserts that len(list(fw.rules)) > 0 right after creation of the firewall before explicitly adding any rules, which implies that fw.rules should indeed return the dummy rules?

It appears to me that fw is None or len(list(fw.rules)) == 0 is the most consistant behaviour we can get :)

Reply:
Yes, I was thinking that all find, list methods etc. would always transparently drop the dummy rule before returning values. You're right, the asserts len(list(fw.rules)) > 0 rule is wrong. It was put in place because aws and openstack both add some default rules when a firewall is created. I think the default is something like a "deny all" rule or something. If GCE does not do the same, then that test is invalid. As long as our dummy rule is effectively a no-op, I think we should be able to safely drop it?

Reply:
Thanks. I deleted the assertion of len(list(fw.rules)) > 0 after a firewall creation and changed the dummy rule behaviour accordingly.